### PR TITLE
tests: enable build on appveyor -- windows continuous integration

### PR DESCRIPTION
**Pull request by fsareshwala**

### CODE REVIEW

File: .appveyor.yml [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+cache:
+    - '%CYG_CACHE%'
+
+init:
+    - git config --global core.autocrlf input
+
+install:
+    - ps: 'Start-FileDownload "http://cygwin.com/setup-x86.exe" -FileName "setup-x86.exe"'
+    - 'setup-x86.exe --quiet-mode --no-shortcuts --only-site --root "%CYG_ROOT%" --site "%CYG_MIRROR%" --local-package-dir "%CYG_CACHE%" --packages gcc-g++,autoconf,automake,make,cmake,libboost-devel,openssl-devel,bzip2,python,git,libstdc++6-devel,libkrb5-devel > NUL 2>&1'
```
Comment (mckurt) [inline .appveyor.yml]: Can we continue this command in next line? It's too long. We can also define packages (gcc-g++,autoconf,automake,...) in an environment variable at top and use that variable.

File: .appveyor.yml [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    - '%CYG_BASH% -lc "cygcheck -dc cygwin"'
+
+build_script:
+    - '%CYG_BASH% -lc "cd $APPVEYOR_BUILD_FOLDER && make gtest tarball"'
+
+# vim: set tw=0:
```
Comment (mckurt) [inline .appveyor.yml]: Is this line still needed?

File: src/cc/tests/CMakeLists.txt [hunk 3]
```diff
@@ -58,7 +58,6 @@ foreach(test_dir ${test_dirs})
 endforeach()
 
 set(test_binary test.t)
-message(STATUS "Test sources: ${test_sources}")
```
Comment (mckurt) [inline src/cc/tests/CMakeLists.txt]: nit: Is there any benefit of limiting the scope of a pull request to changes on the same component? If yes, this commit and a few others should be placed in a separate pull request.

File: src/cc/tests/integtest_main.cc [hunk 4]
```diff
@@ -14,13 +15,31 @@
...
-// TODO(fsareshwala): catch SIGINT and remove kill all environments and remove
-// all files
-
 using namespace KFS::Test;
 using namespace std;
 
 const string QFSTestUtils::kTestHome = "/tmp/qfs";
+static const int kNumChunkServers = 3;
+
+void
```
Comment (mckurt) [inline src/cc/tests/integtest_main.cc]: Can we add  a comment here to clarify why we need this?

File: src/cc/tests/integtest_main.cc [hunk 5]
```diff
@@ -14,13 +15,31 @@
...
+{
+    cout << "Caught signal: " << strsignal(signum) << endl;
+
+    switch (signum) {
+        case SIGSEGV:
+        case SIGINT:
+            cout << "Terminating environments..." << endl;
+            QFSTest::sMetaserver->TearDown();
+            for (unsigned int i = 0; i < kNumChunkServers; i++) {
+                QFSTest::sChunkservers[i]->TearDown();
```
Comment (mckurt) [inline src/cc/tests/integtest_main.cc]: This would be cleaner if we had a static function in QFSTest which calls TearDown() both on sMetaserver and each element in sChunkservers? That function would replace all the lines here.

File: src/cc/common/StrUtils.cc [hunk 6]
```diff
@@ -6,7 +6,7 @@
 namespace KFS {
 
 void
-StrUtils::str_printf_helper(string* dest, const char* format,
+StrUtils::printf_helper(string* dest, const char* format,
```
Comment (mckurt) [inline src/cc/common/StrUtils.cc]: Based on the recent e-mail exchange: should "_" characters be removed? also, should the function names follow camel case?

File: src/cc/tests/integtest.cc [hunk 7]
```diff
@@ -83,23 +93,33 @@ QFSTest::TearDown()
...
 }
 
-string
-QFSTestUtils::WriteTempFile(const string& data)
+int
+QFSTestUtils::CreateTempFile(string* path)
 {
      string filename = QFSTestUtils::kTestHome + "/tmp_file.XXXXXX";
      int fd = mkstemp(const_cast<char*>(filename.c_str()));
      EXPECT_NE(-1, fd) << "Could not mkstep: " << strerror(errno);
```
Comment (mckurt) [inline src/cc/tests/integtest.cc]: Is this a common usage? "EXPECT_NE(...) << ... << ..."
Ignore the comment if it is.

File: src/cc/tests/integtest.h [hunk 8]
```diff
@@ -68,6 +68,15 @@ class QFSTestEnvironment : public ::testing::Environment
     virtual bool Stop() { return true; }
 
     /**
+     * mLogFile is the path to the logfile which contains the output of the
+     * service simulated by this instance of QFSTestEnvironment. These logfiles
+     * are available for review by developers when running tests. On continuous
+     * integration platforms, where inspection is harder, these logfiles are
+     * emitted onto stdout.
+     */
+    string mLogFile;
```
Comment (mckurt) [inline src/cc/tests/integtest.h]: I think I'm missing something. How do we tell chunkservers and metaservers to write into created temporary log file?

**Comment by mckurt** [comment]
There are too many commits modifying different components. Can we group them in separate requests, if you as well see benefit in doing that?

**Comment by mckurt** [comment]
Also, why is windows build still failing?

File: .appveyor.yml (hunk 1 above)
Comment (fsareshwala) [inline .appveyor.yml]: I shortened the line a bit by using the environment variable strategy you suggested.

File: .appveyor.yml (hunk 2 above)
Comment (fsareshwala) [inline .appveyor.yml]: This is a [vim modeline](http://vim.wikia.com/wiki/Modeline_magic). I've removed it but it's useful sometimes to have it if you always find yourself setting the same settings on the same file. Now that the line is shorter through your suggestion above, we don't need this anymore.

File: src/cc/tests/CMakeLists.txt (hunk 3 above)
Comment (fsareshwala) [inline src/cc/tests/CMakeLists.txt]: Yes, actually, there is a lot of benefit to having one pull request per component. I should start doing that. I'm sorry. I'll update this pull request to reflect that and open a few more as well.

File: src/cc/tests/integtest_main.cc (hunk 4 above)
Comment (fsareshwala) [inline src/cc/tests/integtest_main.cc]: Sure, I've added a comment.

File: src/cc/tests/integtest_main.cc (hunk 5 above)
Comment (fsareshwala) [inline src/cc/tests/integtest_main.cc]: Great idea. That's been done.

File: src/cc/common/StrUtils.cc (hunk 6 above)
Comment (fsareshwala) [inline src/cc/common/StrUtils.cc]: I've actually removed this entire class from the codebase now.

File: src/cc/tests/integtest.cc (hunk 7 above)
Comment (fsareshwala) [inline src/cc/tests/integtest.cc]: Yep. It's common to provide custom failure messages like this with Google Test. See the [documentation](https://github.com/google/googletest/blob/master/googletest/docs/V1_7_Primer.md#assertions) for more information. It's useful sometimes to have more context on why a failure occurred for debugging and whatnot.

File: src/cc/tests/integtest.h (hunk 8 above)
Comment (fsareshwala) [inline src/cc/tests/integtest.h]: The chunkservers and metaserver already write their logs out to stderr. In production, we redirect that using daemontools' multilog. Here, we're redirecting the process stderr to the log file. That's done in the `dup2` call inside each environment.

**Comment by fsareshwala** [comment]
I agree I should split this pull request up into multiple smaller ones. I'll do that and keep this one to be just the windows continuous integration change.
